Add import use statements section (#98)
WordPress/wpcs-docs: wordpress-coding-standards/php.md

}

/**
 * Global namespace.
 */
namespace {
    // Code.
}
```

_There is currently no timeline for introducing namespaces to WordPress Core._ 

The use of namespaces in plugins and themes is strongly encouraged. It is a great way to prefix a lot of your code to prevent naming conflicts with other plugins, themes and/or WordPress Core.

Please do make sure you use a unique and long enough namespace prefix to actually prevent conflicts. Generally speaking, using a namespace prefix along the lines of Vendor\Packagename is a good idea.

[warning]
Using `wp` or `WordPress` as a prefix is not allowed.
[/warning]

Namespacing has no effect on variables declared outside of functions/classes, constants declared with `define()` or non-PHP native constructs, like the hook names as used in WordPress.
Those still need to be prefixed individually.

### Using import `use` statements

Using import `use` statements allows you to refer to an external constants, functions, classes, interfaces, and namespaces.

Import `use` statements should be at the top of the file and follow the (optional) `namespace` declaration. They should follow a specific order based on the **type** of the import:

> 1. `use` statements for classes, interfaces a traits

Reviewer: ```suggestion
1. `use` statements for namespaces, classes, interfaces a traits
```

2. `use` statements for functions
3. `use` statements for constants

> When using aliases, make sure the aliases follow the WordPress naming convention, and that you don't set the same name for alias as for the original import statement (this is not allowed: `use My\Project\Class_Name as Class_Name`). The following examples will showcase the correct and incorrect usage of import `use` statements regarding things like spacing, groupings, leading backslashes, etc.

Reviewer: ```suggestion
Aliases can be used to prevent name collisions (two classes in different namespaces using the same class name).
When using aliases, make sure the aliases follow the WordPress naming convention and are unique.

The following examples showcase the correct and incorrect usage of import `use` statements regarding things like spacing, groupings, leading backslashes, etc.
```

I think we don't need to spell out the bit about using the same name (but it should be flagged by a sniff at some point).

> Valid code example:

Reviewer: ```suggestion
Correct:
```

> ```php
> <?php
> /**
>  * File doc block.
>  */
> 
> namespace WordPress\subnamespace;

Reviewer: ```suggestion
```php
namespace WordPress\subnamespace;
```

use Some\Namespace\Class_A;
use Some\Namespace\Class_C as Aliased_Class_C;
use Some\Namespace\{
    Class_D,
    Class_E as Aliased_Class_E,
}

use function Some\Namespace\function_a;
use function Some\Namespace\function_b as aliased_function;

use const Some\Namespace\CONSTANT_A;
use const Some\Namespace\CONSTANT_D as ALIASED_CONSTANT;

// Rest of the code
```

> Invalid code example:

Reviewer: ```suggestion
Incorrect:
```

> ```php
> <?php
> /**
>  * File doc block.
>  */
>  
> namespace WordPress\subnamespace;

Reviewer: ```suggestion
```php
namespace WordPress\subnamespace;
```

> use Some\Namespace \ /*comment*/ Class_A; // Comment inside an import statement.

Reviewer: I think this one can be left out again (parse error in PHP 8.0+ anyway).

use   const   Some\Namespace\CONSTANT_A; // Extra whitespace after the use statement.
use function Some\Namespace\function_a; // Function import after constant import. 

> use \Some\Namespace\Class_C as aliased_class_c; // Leading backslash shouldn't be used, alias has wrong naming.

Reviewer: ```suggestion
use \Some\Namespace\Class_C as aliased_class_c; // Leading backslash shouldn't be used, alias doesn't comply with naming conventions.
```

> use Some\Namespace\{Class_D, Class_E   as   Aliased_Class_E} // Extra spaces around the alias.

Reviewer: ```suggestion
use Some\Namespace\{Class_D, Class_E   as   Aliased_Class_E} // Extra spaces around the "as" keyword, no spaces on the inside of the braces.
```

Those still need to be prefixed individually.
[/info]

### Using import `use` statements

> Using import `use` statements allows you to refer to an external constants, functions, classes, interfaces, and namespaces.

Reviewer: I'm struggling a bit with this sentence, I think mostly because of the use of "external" (external to what ? the answer is "the current namespace", but this will be unclear to people who are not (yet) used to working with namespaces).

Maybe we should use the definition used in the PHP manual, and link to the manual page ?

> The ability to refer to an external fully qualified name with an alias ...

Ref: https://www.php.net/manual/en/language.namespaces.importing.php

Oh and enums and traits should probably also be mentioned (if we want to be complete - I've send a [PR to the PHP docs](https://github.com/php/doc-en/pull/1748) to update the manual page for that too).

Author: Does this sound better?

> One of the important features of namespaces is the ability to refer to an external fully qualified name using an alias (which is also referred to as importing).
>
> PHP can alias(/import) constants, functions, classes, interfaces, traits, enums and namespaces. For more information about importing check the [PHP manual](https://www.php.net/manual/en/language.namespaces.importing.php).

Reviewer 2: > Using import `use` statements allows you to refer to constants, functions, classes, interfaces, and namespaces that live outside of the current namespace.

?

Author: @GaryJones Does this mean that you agree with my proposal above, or should I just remove the alias from the paragraphs and replace it with importing?

Reviewer 2: It means I think my suggestion should be considered as a replacement for what is currently there 😄 

```suggestion
Using import `use` statements allows you to refer to constants, functions, classes, interfaces, and namespaces that live outside of the current namespace.
```

Author: I'll add enums and traits to the list and your suggestion, thanks! 🙂

Import `use` statements should be at the top of the file and follow the (optional) `namespace` declaration. They should follow a specific order based on the **type** of the import:

> 1. `use` statements for namespaces, classes, interfaces a traits

Reviewer: ```suggestion
1. `use` statements for namespaces, classes, interfaces, traits and enums
```

2. `use` statements for functions
3. `use` statements for constants

Aliases can be used to prevent name collisions (two classes in different namespaces using the same class name).
When using aliases, make sure the aliases follow the WordPress naming convention and are unique.

The following examples showcase the correct and incorrect usage of import `use` statements regarding things like spacing, groupings, leading backslashes, etc.

> Correct:

Reviewer: Regarding all code samples - I realize these are copied from the Make post, but.... the examples are fine as theoretical examples, but not as real code examples as `namespace` is a reserved keyword and couldn't be used within a namespace name prior to PHP 8.0: https://3v4l.org/CNaOi

Replacing `Some\Namespace\` with `Some\Name` should be fine -  or possibly better `Prefix\SubPackage\` ?

Author: Totally didn't think about that 🙈 My bad. I'll replace them with some better naming 👍🏼

```php

> namespace WordPress\subnamespace;

Reviewer: ```suggestion
namespace WordPress\Subnamespace;
```

Let's comply with the namespace naming rules we added in the previous PR.... (also for the incorrect example).

use Some\Namespace\Class_A;
use Some\Namespace\Class_C as Aliased_Class_C;
use Some\Namespace\{
    Class_D,
    Class_E as Aliased_Class_E,
}

use function Some\Namespace\function_a;
use function Some\Namespace\function_b as aliased_function;

use const Some\Namespace\CONSTANT_A;
use const Some\Namespace\CONSTANT_D as ALIASED_CONSTANT;

// Rest of the code
```

Incorrect:

```php

> namespace WordPress\subnamespace;

Reviewer 2: Is this namespace line purposely incorrect (capitalization)? Either it needs a comment to say so, or it should be corrected, since the focus in this block should be on the incorrectness of the `use` statements.

Author: I will replace these namespaces, especially since we said that people shouldn't use `WordPress` or `wp` as a namespace.

> use   const   Some\Namespace\CONSTANT_A; // Extra whitespace after the use statement.

Reviewer: ```suggestion
use   const   Some\Namespace\CONSTANT_A; // Superfluous whitespace after the "use" and the "const" keywords.
```

use function Some\Namespace\function_a; // Function import after constant import. 
use \Some\Namespace\Class_C as aliased_class_c; // Leading backslash shouldn't be used, alias doesn't comply with naming conventions.

> use Some\Namespace\{Class_D, Class_E   as   Aliased_Class_E} // Extra spaces around the "as" keyword, no spaces on the inside of the braces.

Reviewer: ```suggestion
use Some\Namespace\{Class_D, Class_E   as   Aliased_Class_E} // Extra spaces around the "as" keyword, incorrect (too little) whitespace use inside the braces.
```

Just re-read the Make post - we explicitly said that group use statements should have the "grouped" parts on new lines, so the "no spaces on the inside of the braces" needs to be rephrased. Not sure if my current suggestion covers it well enough, but probably a step in the right direction.

Author: Maybe:

> Extra spaces around the "as" keyword, incorrect whitespace use inside the brace opener and closer.

> use Vendor\Package\{ function function_a, function function_b,
>      Class_C,
>         const CONSTANT_NAME}; // Combining different types of imports in one use statement.

Reviewer: ```suggestion
use Vendor\Package\{ function function_a, function function_b,
     Class_C,
        const CONSTANT_NAME}; // Combining different types of imports in one use statement, incorrect whitespace use within group use statement.
```

class Foo {
    // Code.
}

use const \Some\Namespace\CONSTANT_D as Aliased_constant; // Import after class definition, leading backslash, naming violation.

> use function Some\Namespace\function_b as Aliased_Function; // Import after class definition, naming violation.

Reviewer: "naming violation" => "naming conventions violation" (x2)

// Rest of the code
```

[alert]
Import `use` statements have no effect on dynamic class, function or constant names. 
Group `use` statements are available from PHP 7.0, and trailing commas in group `use` statements are available from PHP 7.2. 
[/alert]

[info]

> Note that, unless you have some sort of [autoloading](https://www.php.net/manual/en/language.oop5.autoload.php) in place, the `use` statement won't automatically load whatever is being imported. You'll either need to set up autoloading, or manually load the file containing the class/function/constant using a `require/import` statement.

Reviewer: ```suggestion
Note that, unless you have implemented [autoloading](https://www.php.net/manual/en/language.oop5.autoload.php), the `use` statement won't automatically load whatever is being imported. You'll either need to set up autoloading, or manually load the file containing the class/function/constant using a `require/import` statement.
```

Note: technically this is still incorrect - `use` statements don't trigger autoloading. Autoloading will only kick in when the actual class/function etc is being _used_. The `use` statements are used by PHP for the name resolution, after which the autoloading may kick in.

```php

use Some\ClassName;

class Foo {
    public function bar() {
        $a = new ClassName(); // <- This is where the autoloading takes place and only if the Foo::bar() function is actually called during the request.
    }
}
```

Author: > Note that, unless you have implemented [autoloading](https://www.php.net/manual/en/language.oop5.autoload.php), the `use` statement won't automatically load whatever is being imported. You'll either need to set up autoloading or load the file containing the class/function/constant using a `require/import` statement, for the aliased constructs to be loaded when used.

Better (note the addition to be loaded when used at the end)?

[/info]

**Note about WordPres Core usage**

While import `use` statements can already be used in WordPress Core, it is, for the moment, **strongly discouraged**.

Import `use` statements are most useful when combined with namespaces and a class autoloading implementation.

> As neither of these are currently in place for WordPress Core and discussions about this are ongoing, holding off on adding import `use` statements to WordPress Core is the sane choice for now.

Reviewer 2: ```suggestion
As neither of these are currently in place for WordPress Core and discussions about this are ongoing, holding off on adding import `use` statements to WordPress Core is the sensible choice for now.
```

Reviewer: Oh, good one! 💯
